Insert multiple rows at once in to_sql

maxispeicher opened the pull request:
*Issue #599:*

*Description of changes:*
Changing the generation of the `INSERT` sql query to include `chunksize` number of rows per query.

Additionally the RDS related tests now use a `pytest` fixture for the connections wherever possible, to ensure closing of the connection.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

igorborgest reviewed (commented):
Everything looks great @maxispeicher.

Do you see any downside for keeping the default chunksize bigger then one?

jaidisido reviewed (approved):
LGTM

maxispeicher commented:
The only downside I see, is easier debugging when a row cannot be inserted. But I also think something like 100 would probably be best for the majority of users.
I initially kept it at 1 for ensuring the same behavior as it is right now.

igorborgest commented:
@maxispeicher Yeah, I was thinking the same.

We are planning to add this enhancement on version `2.6.0` so this behavior change is acceptable .

Do you mind to update the PR and add a higher default value? I have no data points to guide us here but I think 100 is a good starting point. We can tune it later based on customer feedback or further tests.

maxispeicher commented:
I did some tests regarding `chunksize` with a local `postgresql` docker container (so IO is negligible) with the following script:
```python
import time

import pandas as pd
import pg8000

import awswrangler as wr

con = pg8000.connect(user="postgres", password="mysecretpassword", port=5432, host="127.0.0.1")
str_col_values = ["foo", "bar", "baz"]
number_of_rows = 100_000
df = pd.DataFrame({
    "c0": [i for i in range(number_of_rows)],
    "c1": [str_col_values[i % 3] for i in range(number_of_rows)],
    "c2": [pd.Timestamp.now() + pd.Timedelta(f"{i}s") for i in range(number_of_rows)]
})

chunksizes = [1, 10, 50, 100, 250, 500, 1000]
for chunksize in chunksizes:
    print(f"Starting insertion for {chunksize=}")

    start_time = time.perf_counter()
    wr.postgresql.to_sql(df=df, con=con, schema="public", table=f"test_table_{chunksize}", chunksize=chunksize)
    con.commit()

    elapsed_time = time.perf_counter() - start_time
    print(f"Inserting with {chunksize=} took {elapsed_time} seconds.")

con.close()
```
with the following results:
```
Inserting with chunksize=1 took 222.9119826 seconds.
Inserting with chunksize=10 took 24.904498599999982 seconds.
Inserting with chunksize=50 took 7.0587190999999905 seconds.
Inserting with chunksize=100 took 4.854258099999981 seconds.
Inserting with chunksize=250 took 3.382871000000023 seconds.
Inserting with chunksize=500 took 3.0053217000000245 seconds.
Inserting with chunksize=1000 took 2.80447879999997 seconds.
```

I would therefore set it to 200 as a first tradeoff between speed improvement and chunksize.
Updated commit will follow soon.

igorborgest reviewed (approved):
Awesome!

maxispeicher commented:
I want to check one possible improvement tomorrow morning therefore just conerting to draft :)

koiker commented on awswrangler/sqlserver.py:
```diff
@@ -291,6 +291,7 @@ def to_sql(
     dtype: Optional[Dict[str, str]] = None,
     varchar_lengths: Optional[Dict[str, int]] = None,
     use_column_names: bool = False,
+    chunksize: int = 200,
```
Use the config class to set this value and make available to users or other modules to change this in code.

maxispeicher replied:
Done 👍

maxispeicher commented:
Changes since the last review:

- Generating the placeholders and parameters for the sql queries with a generator instead of all at once (minimal performance impact but memory-saving for larger dfs)
- Adding chunksize to config
- Fixing issue with generation of docs for `sqlserver` module
